File: libheifer/tests/color.c
/* SPDX-License-Identifier: LGPL-3.0-or-later */
/* Independent original-header client. Never compare padding or indeterminate
 * decoded coordinates returned by the upstream NCLX allocator. */
#include <libheif/heif.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
static void error(heif_error e) { printf(" %d:%d:%s", e.code, e.subcode, e.message ? e.message : "NULL"); }
static void f32(float f) { uint32_t x; memcpy(&x,&f,4); printf(" %08" PRIx32,x); }
static void f64(double d) { uint64_t x; memcpy(&x,&d,8); printf(" %016" PRIx64,x); }
static void profile(heif_color_profile_nclx* p, int decoded) {
  printf(" %u:%d:%d:%d:%u",p->version,p->color_primaries,p->transfer_characteristics,p->matrix_coefficients,p->full_range_flag);
  if (decoded) {
    f32(p->color_primary_red_x); f32(p->color_primary_red_y);
    f32(p->color_primary_green_x); f32(p->color_primary_green_y);
    f32(p->color_primary_blue_x); f32(p->color_primary_blue_y);
    f32(p->color_primary_white_x); f32(p->color_primary_white_y);
  }
}
static void get_nclx(heif_image* image) {
  heif_color_profile_nclx* p=(void*)(uintptr_t)0x1234;
  heif_error e=heif_image_get_nclx_color_profile(image,&p);
  error(e); printf(" out=%d:%d",p==NULL,p==(void*)(uintptr_t)0x1234);
  if (!e.code) { profile(p,1); heif_nclx_color_profile_free(p); }
}
static void options(heif_color_conversion_options_ext* p) {
  printf(" %u:%d:%u:%u:%u:%u:%u:%u:%u",p->version,p->alpha_composition_mode,p->background_red,p->background_green,p->background_blue,p->secondary_background_red,p->secondary_background_green,p->secondary_background_blue,p->checkerboard_square_size);
}
static void mastering(heif_mastering_display_colour_volume* p) {
  for(int j=0;j<3;j++) printf(" %u:%u",p->display_primaries_x[j],p->display_primaries_y[j]);
  printf(" %u:%u:%" PRIu32 ":%" PRIu32,p->white_point_x,p->white_point_y,p->max_display_mastering_luminance,p->min_display_mastering_luminance);
}
int main(void) {
  heif_color_conversion_options opts; memset(&opts,0xa5,sizeof(opts));
  heif_color_conversion_options_set_defaults(&opts);
  printf("defaults %u %d %d %u",opts.version,opts.preferred_chroma_downsampling_algorithm,opts.preferred_chroma_upsampling_algorithm,opts.only_use_preferred_chroma_algorithm);
  heif_color_conversion_options_ext* ext=heif_color_conversion_options_ext_alloc(); if(!ext) return 1;
  options(ext); heif_color_conversion_options_ext_free(ext); heif_color_conversion_options_ext_free(NULL);
  heif_color_profile_nclx* allocated=heif_nclx_color_profile_alloc(); if(!allocated) return 2;
  profile(allocated,0); heif_nclx_color_profile_free(allocated); heif_nclx_color_profile_free(NULL); puts("");
  for(unsigned dv=0;dv<256;dv++) for(unsigned sv=0;sv<256;sv++) {
    heif_color_conversion_options_ext dst={0},src={0}; dst.version=dv; src.version=sv;
    src.alpha_composition_mode=2; src.background_red=17; src.background_green=123; src.background_blue=65535;
    src.secondary_background_red=65534; src.secondary_background_green=65533; src.secondary_background_blue=65532; src.checkerboard_square_size=65432;
    heif_color_conversion_options_ext_copy(&dst,NULL);
    heif_color_conversion_options_ext_copy(&dst,&src);
    heif_color_conversion_options_ext_copy(&dst,&dst);
    printf("options %u:%u",dv,sv); options(&dst); puts("");
  }
  for(unsigned value=0;value<65536;value++) {
    heif_color_profile_nclx p={0}; p.version=9; p.full_range_flag=7; p.color_primary_red_x=0.123f;
    printf("setters %u",value);
    error(heif_nclx_color_profile_set_color_primaries(&p,value));
    error(heif_nclx_color_profile_set_transfer_characteristics(&p,value));
    error(heif_nclx_color_profile_set_matrix_coefficients(&p,value)); profile(&p,1); puts("");
  }
  int vals[]={-1,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,22,65535,65536,65537,65538,0x7fffffff};
  for(unsigned a=0;a<sizeof(vals)/sizeof(vals[0]);a++) for(unsigned b=0;b<sizeof(vals)/sizeof(vals[0]);b++) for(unsigned r=0;r<3;r++) {
    heif_image* image=NULL; if(heif_image_create(1,1,heif_colorspace_RGB,heif_chroma_444,&image).code) return 3;
    printf("nclx %d:%d:%u type=%u",vals[a],vals[b],r,(unsigned)heif_image_get_color_profile_type(image));
    get_nclx(image); error(heif_image_get_nclx_color_profile(image,NULL));
    heif_color_profile_nclx p={0}; p.version=0; p.color_primaries=vals[a]; p.transfer_characteristics=vals[b]; p.matrix_coefficients=(a%2)?vals[b]:2; p.full_range_flag=r?((r==1)?1:255):0;
    error(heif_image_set_nclx_color_profile(image,&p));
    p.color_primaries=99; p.color_primary_red_x=123; /* input is copied */
    printf(" type=%u",(unsigned)heif_image_get_color_profile_type(image)); get_nclx(image);
    heif_image_release(image); puts("");
  }
  const char* types[]={"","a","abc","prof","rICC","nclx","zzzz","profx","p\0of","\xff\xff\xff\xff"};
  unsigned sizes[]={0,1,3,4,255,256,1024};
  for(unsigned t=0;t<sizeof(types)/sizeof(types[0]);t++) for(unsigned s=0;s<sizeof(sizes)/sizeof(sizes[0]);s++) {
    heif_image* image=NULL; if(heif_image_create(1,1,heif_colorspace_RGB,heif_chroma_444,&image).code) return 4;
    printf("raw %u:%u",t,sizes[s]);
    unsigned char data[1024],out[1026]; memset(out,0xa5,sizeof(out));
    error(heif_image_get_raw_color_profile(image,out+1)); error(heif_image_get_raw_color_profile(image,NULL));
    heif_color_profile_nclx p={0}; p.color_primaries=1; p.transfer_characteristics=13; p.matrix_coefficients=6;
    error(heif_image_set_nclx_color_profile(image,&p));
    for(int round=0;round<2;round++) {
      for(unsigned i=0;i<sizeof(data);i++) data[i]=(i*37+round*11)&255;
      error(heif_image_set_raw_color_profile(image,types[t],data,sizes[s])); memset(data,0xa5,sizeof(data));
      printf(" type=%u size=%zu",(unsigned)heif_image_get_color_profile_type(image),heif_image_get_raw_color_profile_size(image));
      error(heif_image_get_raw_color_profile(image,out+1));
      for(unsigned i=0;i<sizes[s]+2;i++) printf("%02x",out[i]);
    }
    get_nclx(image); heif_image_release(image); puts("");
  }
  for(unsigned value=0;value<65536;value++) {
    heif_mastering_display_colour_volume in={0}; heif_decoded_mastering_display_colour_volume out;
    for(unsigned j=0;j<3;j++) { in.display_primaries_x[j]=(value+j)&65535; in.display_primaries_y[j]=(value+65535-j)&65535; }
    in.white_point_x=value; in.white_point_y=value;
    in.max_display_mastering_luminance=value*1703u; in.min_display_mastering_luminance=value;
    printf("decode %u",value); error(heif_mastering_display_colour_volume_decode(&in,&out));
    for(unsigned j=0;j<3;j++) { f32(out.display_primaries_x[j]); f32(out.display_primaries_y[j]); }
    f32(out.white_point_x); f32(out.white_point_y); f64(out.max_display_mastering_luminance); f64(out.min_display_mastering_luminance); puts("");
  }
  uint32_t bounds[]={0,1,4,5,37000,37001,42000,42001,49999,50000,50001,99999999,100000000,100000001,UINT32_MAX};
  for(unsigned a=0;a<sizeof(bounds)/sizeof(bounds[0]);a++) for(unsigned b=0;b<sizeof(bounds)/sizeof(bounds[0]);b++) {
    heif_image* image=NULL; if(heif_image_create(1,1,heif_colorspace_RGB,heif_chroma_444,&image).code) return 5;
    heif_content_light_level light={bounds[a],bounds[b]},lg={123,456};
    heif_ambient_viewing_environment ambient={bounds[a],bounds[b],bounds[a]},ag={123,456,789};
    printf("hdr %u:%u initial=%d:%d:%d:%d:%" PRIu32,a,b,heif_image_has_content_light_level(image),heif_image_has_mastering_display_colour_volume(image),heif_image_has_ambient_viewing_environment(image),heif_image_has_nominal_diffuse_white_luminance(image),heif_image_get_nominal_diffuse_white_luminance(image));
    heif_image_get_content_light_level(image,&lg); printf(" light=%u:%u",lg.max_content_light_level,lg.max_pic_average_light_level);
    int present=heif_image_get_ambient_viewing_environment(image,&ag); printf(" ambient=%d:%" PRIu32 ":%u:%u",present,ag.ambient_illumination,ag.ambient_light_x,ag.ambient_light_y);
    heif_image_set_content_light_level(image,NULL); heif_image_set_mastering_display_colour_volume(image,NULL); heif_image_set_ambient_viewing_environment(image,NULL);
    heif_image_get_content_light_level(image,NULL); printf(" absent=%d",heif_image_get_ambient_viewing_environment(image,NULL));
    heif_image_set_content_light_level(image,&light); heif_image_set_ambient_viewing_environment(image,&ambient); memset(&light,0,sizeof(light)); memset(&ambient,0,sizeof(ambient));
    heif_image_set_nominal_diffuse_white_luminance(image,bounds[a]);
    heif_image_get_content_light_level(image,&lg); present=heif_image_get_ambient_viewing_environment(image,&ag);
    printf(" stored=%d:%d:%d:%" PRIu32 " light=%u:%u ambient=%d:%" PRIu32 ":%u:%u",heif_image_has_content_light_level(image),heif_image_has_ambient_viewing_environment(image),heif_image_has_nominal_diffuse_white_luminance(image),heif_image_get_nominal_diffuse_white_luminance(image),lg.max_content_light_level,lg.max_pic_average_light_level,present,ag.ambient_illumination,ag.ambient_light_x,ag.ambient_light_y);
    heif_mastering_display_colour_volume in={0},got={0}; heif_decoded_mastering_display_colour_volume decoded={0};
    in.display_primaries_x[0]=bounds[a]; in.display_primaries_y[2]=bounds[b]; in.white_point_x=bounds[b]; in.white_point_y=bounds[a]; in.max_display_mastering_luminance=bounds[a]; in.min_display_mastering_luminance=bounds[b];
    heif_image_set_mastering_display_colour_volume(image,&in); memset(&in,0,sizeof(in)); heif_image_get_mastering_display_colour_volume(image,&got);
    printf(" mastering=%d",heif_image_has_mastering_display_colour_volume(image)); mastering(&got);
    error(heif_mastering_display_colour_volume_decode(&got,&decoded)); f64(decoded.max_display_mastering_luminance); f64(decoded.min_display_mastering_luminance);
    error(heif_mastering_display_colour_volume_decode(NULL,&decoded)); error(heif_mastering_display_colour_volume_decode(&got,NULL));
    heif_image_set_content_light_level(image,&light); printf(" reset=%d",heif_image_has_content_light_level(image));
    heif_image_release(image); puts("");
  }
  return 0;
}

File: libheifer/src/color.rs
// SPDX-License-Identifier: LGPL-3.0-or-later
// Compatibility semantics adapted from libheif, Copyright Dirk Farin and contributors.
use crate::error::Error;

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct ColorConversionOptions {
    pub version: u8,
    pub preferred_chroma_downsampling_algorithm: i32,
    pub preferred_chroma_upsampling_algorithm: i32,
    pub only_use_preferred_chroma_algorithm: u8,
}
impl Default for ColorConversionOptions {
    fn default() -> Self {
        // Matches the reference without the optional native libsharpyuv backend.
        Self {
            version: 1,
            preferred_chroma_downsampling_algorithm: 2,
            preferred_chroma_upsampling_algorithm: 2,
            only_use_preferred_chroma_algorithm: 1,
        }
    }
}
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct ColorConversionOptionsExt {
    pub version: u8,
    pub alpha_composition_mode: i32,
    pub background_red: u16,
    pub background_green: u16,
    pub background_blue: u16,
    pub secondary_background_red: u16,
    pub secondary_background_green: u16,
    pub secondary_background_blue: u16,
    pub checkerboard_square_size: u16,
}
impl Default for ColorConversionOptionsExt {
    fn default() -> Self {
        Self {
            version: 1,
            alpha_composition_mode: 0,
            background_red: 0xffff,
            background_green: 0xffff,
            background_blue: 0xffff,
            secondary_background_red: 0xcccc,
            secondary_background_green: 0xcccc,
            secondary_background_blue: 0xcccc,
            checkerboard_square_size: 16,
        }
    }
}
impl ColorConversionOptionsExt {
    pub fn copy_from_versioned(&mut self, src: Self) {
        if self.version.min(src.version) == 1 {
            let version = self.version;
            *self = src;
            self.version = version;
        }
    }
}

#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct NclxProfile {
    pub version: u8,
    pub color_primaries: i32,
    pub transfer_characteristics: i32,
    pub matrix_coefficients: i32,
    pub full_range_flag: u8,
    pub color_primary_red_x: f32,
    pub color_primary_red_y: f32,
    pub color_primary_green_x: f32,
    pub color_primary_green_y: f32,
    pub color_primary_blue_x: f32,
    pub color_primary_blue_y: f32,
    pub color_primary_white_x: f32,
    pub color_primary_white_y: f32,
}

impl Default for NclxProfile {
    fn default() -> Self {
        Self {
            version: 1,
            color_primaries: 1,
            transfer_characteristics: 13,
            matrix_coefficients: 6,
            full_range_flag: 1,
            // Upstream's allocator leaves decoded coordinates unspecified.
            color_primary_red_x: 0.0,
            color_primary_red_y: 0.0,
            color_primary_green_x: 0.0,
            color_primary_green_y: 0.0,
            color_primary_blue_x: 0.0,
            color_primary_blue_y: 0.0,
            color_primary_white_x: 0.0,
            color_primary_white_y: 0.0,
        }
    }
}

pub fn validate_primaries(value: u16) -> Result<(), Error> {
    if matches!(value, 1 | 2 | 4..=12 | 22) {
        Ok(())
    } else {
        Err(Error::new(2, 133, c"Unknown error"))
    }
}
pub fn validate_transfer(value: u16) -> Result<(), Error> {
    if matches!(value, 1 | 2 | 4..=18) {
        Ok(())
    } else {
        Err(Error::new(2, 134, c"Unknown error"))
    }
}
pub fn validate_matrix(value: u16) -> Result<(), Error> {
    if matches!(value, 0..=2 | 4..=14) {
        Ok(())
    } else {
        Err(Error::new(2, 135, c"Unknown error"))
    }
}

/// Stored image metadata is narrowed to 16 bits, matching the container model.
#[derive(Clone, Copy, Debug)]
pub struct Nclx {
    pub primaries: u16,
    pub transfer: u16,
    pub matrix: u16,
    pub full_range: bool,
}

impl Nclx {
    pub fn is_defined(self) -> bool {
        self.primaries != 2 || self.transfer != 2 || self.matrix != 2 || !self.full_range
    }
    pub fn decode(self) -> Result<NclxProfile, Error> {
        validate_primaries(self.primaries)
            .map_err(|_| Error::new(2, 133, c"Invalid input: Unknown NCLX color primaries"))?;
        validate_transfer(self.transfer).map_err(|_| {
            Error::new(
                2,
                134,
                c"Invalid input: Unknown NCLX transfer characteristics",
            )
        })?;
        validate_matrix(self.matrix)
            .map_err(|_| Error::new(2, 135, c"Invalid input: Unknown NCLX matrix coefficients"))?;
        let [gx, gy, bx, by, rx, ry, wx, wy] = match self.primaries {
            1 => [0.300, 0.600, 0.150, 0.060, 0.640, 0.330, 0.3127, 0.3290],
            4 => [0.21, 0.71, 0.14, 0.08, 0.67, 0.33, 0.310, 0.316],
            5 => [0.29, 0.60, 0.15, 0.06, 0.64, 0.33, 0.3127, 0.3290],
            6 | 7 => [0.310, 0.595, 0.155, 0.070, 0.630, 0.340, 0.3127, 0.3290],
            8 => [0.243, 0.692, 0.145, 0.049, 0.681, 0.319, 0.310, 0.316],
            9 => [0.170, 0.797, 0.131, 0.046, 0.708, 0.292, 0.3127, 0.3290],
            10 => [0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.333333, 0.33333],
            11 => [0.265, 0.690, 0.150, 0.060, 0.680, 0.320, 0.314, 0.351],
            12 => [0.265, 0.690, 0.150, 0.060, 0.680, 0.320, 0.3127, 0.3290],
            22 => [0.295, 0.605, 0.155, 0.077, 0.630, 0.340, 0.3127, 0.3290],
            _ => [0.0; 8],
        };
        Ok(NclxProfile {
            version: 1,
            color_primaries: self.primaries.into(),
            transfer_characteristics: self.transfer.into(),
            matrix_coefficients: self.matrix.into(),
            full_range_flag: self.full_range.into(),
            color_primary_red_x: rx,
            color_primary_red_y: ry,
            color_primary_green_x: gx,
            color_primary_green_y: gy,
            color_primary_blue_x: bx,
            color_primary_blue_y: by,
            color_primary_white_x: wx,
            color_primary_white_y: wy,
        })
    }
}

#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct ContentLightLevel {
    pub max_content_light_level: u16,
    pub max_pic_average_light_level: u16,
}
#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct MasteringDisplayColourVolume {
    pub display_primaries_x: [u16; 3],
    pub display_primaries_y: [u16; 3],
    pub white_point_x: u16,
    pub white_point_y: u16,
    pub max_display_mastering_luminance: u32,
    pub min_display_mastering_luminance: u32,
}
#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct DecodedMasteringDisplayColourVolume {
    pub display_primaries_x: [f32; 3],
    pub display_primaries_y: [f32; 3],
    pub white_point_x: f32,
    pub white_point_y: f32,
    pub max_display_mastering_luminance: f64,
    pub min_display_mastering_luminance: f64,
}
impl MasteringDisplayColourVolume {
    pub fn decode(self) -> DecodedMasteringDisplayColourVolume {
        fn coordinate(value: u16, max: u16) -> f32 {
            if (5..=max).contains(&value) {
                (f64::from(value) * 0.00002) as f32
            } else {
                0.0
            }
        }
        fn luminance(value: u32, min: u32, max: u32) -> f64 {
            if (min..=max).contains(&value) {
                f64::from(value) * 0.0001
            } else {
                0.0
            }
        }
        DecodedMasteringDisplayColourVolume {
            display_primaries_x: self.display_primaries_x.map(|v| coordinate(v, 37000)),
            display_primaries_y: self.display_primaries_y.map(|v| coordinate(v, 42000)),
            white_point_x: coordinate(self.white_point_x, 37000),
            white_point_y: coordinate(self.white_point_y, 42000),
            max_display_mastering_luminance: luminance(
                self.max_display_mastering_luminance,
                50000,
                100000000,
            ),
            min_display_mastering_luminance: luminance(
                self.min_display_mastering_luminance,
                1,
                50000,
            ),
        }
    }
}
#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct AmbientViewingEnvironment {
    pub ambient_illumination: u32,
    pub ambient_light_x: u16,
    pub ambient_light_y: u16,
}

#[derive(Debug)]
pub struct RawProfile {
    pub profile_type: u32,
    pub data: Vec<u8>,
}
#[derive(Debug, Default)]
pub struct ColorMetadata {
    pub raw: Option<RawProfile>,
    pub nclx: Option<Nclx>,
    pub content_light: ContentLightLevel,
    pub mastering: Option<MasteringDisplayColourVolume>,
    pub ambient: Option<AmbientViewingEnvironment>,
    pub diffuse_white: Option<u32>,
}
impl ColorMetadata {
    pub fn profile_type(&self) -> u32 {
        self.raw.as_ref().map_or_else(
            || {
                if self.nclx.is_some_and(Nclx::is_defined) {
                    u32::from_be_bytes(*b"nclx")
                } else {
                    0
                }
            },
            |r| r.profile_type,
        )
    }
    pub fn set_raw(&mut self, profile_type: u32, data: &[u8]) -> Result<(), Error> {
        let mut copy = Vec::new();
        copy.try_reserve_exact(data.len())
            .map_err(|_| Error::ALLOCATION)?;
        copy.extend_from_slice(data);
        self.raw = Some(RawProfile {
            profile_type,
            data: copy,
        });
        Ok(())
    }
    pub fn has_content_light(&self) -> bool {
        self.content_light.max_content_light_level != 0
            || self.content_light.max_pic_average_light_level != 0
    }
}
